proposals/list-patterns-on-enumerables-2.md

```diff
@@ -0,0 +1,91 @@
+# List patterns on enumerables
+
+Champion issue: <https://github.com/dotnet/csharplang/issues/9005>
+
+## Summary
+
+Extends list patterns to be able to be used with enumerables that are not countable or indexable. `items.Where(...) is [var singleItem]`, or `is []`, or `is [p1, .., p2]`.
+
+The pattern will be evaluated without multiple enumeration. The slice pattern `..` is supported, but only without a subpattern.
+
+## Motivation
+
+[LDM 2023-10-09](https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-10-09.md#list-patterns-on-enumerables) set the following direction:
+
+> This is follow-up work from C# 11 that we did not have time in C# 12 to invest in. We intend to continue the work here now; collection expressions supporting more than just indexable and countable types show where our list pattern support falls short.
+
+One example of where gap is noticed is when using LINQ methods. `.Where(...)` is the type of thing which is common to insert after a collection, but when you insert this in `items is [var item]` or similar, the list pattern is no longer permitted. This puts you in an awkward spot where a lot of rewriting is necessary. There's no built-in helper that recovers the behavior of `is [var item]`. With more complex list patterns, it only gets worse from there.
+
+## Detailed design
+
+Any list pattern will be supported for an enumerable type (a type supported by `foreach`) if the same pattern would be supported by a type that is countable and indexable, but not sliceable (even if the enumerable is sliceable). Thus, for the enumerable types gaining support through this proposal, it will be an error for a slice pattern to contain a subpattern.
+
+The type being matched against for each element pattern inside the list pattern will be determined the same way the iteration variable type is inferred with the `foreach` statement. For the slice pattern `..` without a subpattern, the type that it is matching against is unspecified.
+
+Async enumerables are not supported. So far in the language, consumption of async enumerables requires the `await` keyword which highlights the point where execution may be suspended.
+
+No new syntax is involved in this proposal.
+
+### Design rationale
+
+#### No multiple enumeration
+
+Enumerables cannot be assumed to represent a materialized collection. An enumerable may represent an in-memory collection or a generated sequence, but it may also represent a remote query or an iterator method. As such, an enumerable may return different results on each enumeration, and enumeration may have side effects. Multiple enumeration is considered both a performance smell and a correctness issue in cases where the enumerable is not known to be a materialized collection or a trivially generated, guaranteed-stable sequence. The .NET SDK and other popular tools produce warnings for multiple enumeration of the same enumerable.
```

Comment:
Needs consideration: is the no-multiple-enumeration guarantee even possible when switching on the enumerable?

Reply:
I don't think so. I also don't think any of these scenarios should be expected to reduce multiple enumeration compared to each other:
```cs
enumerable is [1] or [2, 3]
enumerable is [1] || enumerable is [2, 3]
enumerable.SequenceEqual([1]) || enumerable.SequenceEqual([2, 3])
```
The existing analyzers that warn about multiple enumeration in the last case can be augmented to see it happening in the prior two cases as well.

Reply:
The compiler _may_ be able to guarantee no multiple enumeration though. And then maybe `is [1] or [1, 2]` may not have to enumerate multiple times, and maybe this is worth guaranteeing.

Reply:
```csharp
enumerable is [1] || enumerable is [2, 3]
```

I think that must result in multiple enumerations because the `||` is not pattern matching. Basically it's two independent pattern matching expressions each of which has single enumeration of `enumerable`. 

```csharp
enumerable is [1] or [2, 3]
```

This though is a case where we'd want to avoid multiple enumerations.

Reply:
> The compiler may be able to guarantee no multiple enumeration though. 

Think that would be important for us to support list patterns on `IEnumerable<T>`. Basically within a single pattern expression or `switch` there should be only a single enumeration of the enumerable. Once you break out of the pattern, with say an `||`, then all bets are off.